File: debug/src/main/java/com/sofar/debug/performance/monitor/MemoryMonitor.java
package com.sofar.debug.performance.monitor;

import android.app.ActivityManager;
import android.content.Context;
import android.os.Build;
import android.os.Debug;
import android.os.Process;
import android.util.Log;

import com.sofar.debug.performance.PerformanceMonitorManager;

public class MemoryMonitor extends AbsMonitor {

  private static final int SAMPLING_TIME = 1000;
  private static final String TAG = "MemoryMonitor";

  private float mLastMemory;
  ActivityManager activityManager;

  private Runnable runnable = new Runnable() {
    @Override
    public void run() {
      mLastMemory = getMemoryData();
      record();
      handler.postDelayed(this, SAMPLING_TIME);
    }
  };

  public MemoryMonitor() {
    activityManager = (ActivityManager) appContext.getSystemService(Context.ACTIVITY_SERVICE);
  }

  @Override
  protected void onStart() {
    handler.postDelayed(runnable, SAMPLING_TIME);
  }

  @Override
  protected void onStop() {
    handler.removeCallbacks(runnable);
  }

  private void record() {
    ActivityManager.MemoryInfo info = new ActivityManager.MemoryInfo();
    activityManager.getMemoryInfo(info);
    float total = info.totalMem / 1024 / 1024;
    float avail = info.availMem / 1024 / 1024;
    Log.d(TAG, "当前使用内存=" + mLastMemory + "MB" + " 总内存=" + total + "MB" + " 可用内存=" + avail + "MB");
    PerformanceMonitorManager.get().memoryCallback(mLastMemory);
  }

  private float getMemoryData() {
    float mem = 0.0F;
    try {
      Debug.MemoryInfo memInfo = null;
      if (Build.VERSION.SDK_INT > Build.VERSION_CODES.P) {
        memInfo = new Debug.MemoryInfo();
        Debug.getMemoryInfo(memInfo);
      } else {
        Debug.MemoryInfo[] memInfos =
          activityManager.getProcessMemoryInfo(new int[]{Process.myPid()});
        if (memInfos != null && memInfos.length > 0) {
          memInfo = memInfos[0];
        }
      }

      int totalPss = memInfo.getTotalPss();
      if (totalPss >= 0) {
        mem = totalPss / 1024.0F;
      }
    } catch (Exception e) {
      e.printStackTrace();
    }
    return mem;
  }
}

File: tool/profiler/src/main/java/com/sofar/profiler/memory/MemoryMonitor.kt
package com.sofar.profiler.memory

import android.os.Build
import android.os.Debug
import android.util.Log
import com.sofar.profiler.AbsMonitor
import com.sofar.profiler.MonitorManager
import com.sofar.profiler.MonitorType
import com.sofar.profiler.memory.model.MemoryInfo
import java.lang.Exception

class MemoryMonitor : AbsMonitor() {

  private var tag = "MemoryMonitor"
  private lateinit var memInfo: MemoryInfo

  private val runnable: Runnable = object : Runnable {
    override fun run() {
      record()
      handler.postDelayed(this, pollInterval())
    }
  }

  override fun onStart() {
    handler.removeCallbacks(runnable)
    handler.postDelayed(runnable, pollInterval())
  }

  override fun onStop() {
    handler.removeCallbacks(runnable)
  }

  override fun type(): MonitorType {
    return MonitorType.MEMORY
  }

  fun getMemoryInfo(): String {
    return memInfo.toString()
  }

  private fun record() {
    try {
      pssMemory()
      Log.d(tag, memInfo.toString())
      MonitorManager.memoryCallback(memInfo)
    } catch (e: Exception) {
      Log.d(tag, e.toString())
    }
  }

  private fun pssMemory() {
    var mem = Debug.MemoryInfo()
    Debug.getMemoryInfo(mem)

    var pssTotalK = mem.totalPss
    var pssJavaK = -1
    var pssNativeK = -1
    var pssGraphicK = -1
    var pssStackK = -1
    var pssCodeK = -1

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      mem.memoryStats.apply {
        fun Map<String, String>.getInt(key: String) = get(key)?.toInt() ?: -1
        pssJavaK = getInt("summary.java-heap")
        pssNativeK = getInt("summary.native-heap")
        pssCodeK = getInt("summary.code")
        pssStackK = getInt("summary.stack")
        pssGraphicK = getInt("summary.graphics")
      }
    } else {
      mem.apply {
        pssJavaK = dalvikPrivateDirty
        pssNativeK = nativePrivateDirty
      }
    }

    memInfo = MemoryInfo(pssTotalK, pssJavaK, pssNativeK, pssGraphicK, pssStackK, pssCodeK)
  }

}